content/en/blog/_posts/2024-02-29-cri-o-seccomp-oci-artifacts.md

```diff
@@ -0,0 +1,421 @@
+INFO[…] Trying to read profile from: /var/lib/containers/storage/overlay/c15f3eb1efd9abe0ae973e050c374988bba52c45a53f5c3cd80419a865013fa6/merged/seccomp.json  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Retrieved OCI artifact seccomp profile of len: 18853  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Created container 116a316cd9a11fe861dd04c43b94f45046d1ff37e2ed05a4e4194fcaab29ee63: default/pod/container  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+```
+
+```
+> export CONTAINER_ID=$(sudo crictl ps --name container -q)
+
+> sudo crictl inspect $CONTAINER_ID | jq .info.runtimeSpec.linux.seccomp | head
+{
+  "defaultAction": "SCMP_ACT_ERRNO",
+  "defaultErrnoRet": 38,
+  "architectures": [
+    "SCMP_ARCH_X86_64",
+    "SCMP_ARCH_X86",
+    "SCMP_ARCH_X32"
+  ],
+  "syscalls": [
+    {
+```
+
+The cool thing about this is that users can now create seccomp profiles for
+specific container images and store them side by side in the same registry.
+Linking the images to the profiles provides a great flexibility to maintain both
+over the whole application's life cycle.
+
+### Pushing profiles using ORAS
+
+The actual creation of the profiles requires a bit more prework, but I have the
+hope that future tools will simplify the overall process. Right now, the
+container registry needs to be [ORAS
+compatible](https://oras.land/docs/compatible_oci_registries/#registries-supporting-oci-artifacts),
+which is also (partially) the case for [Quay.io](https://quay.io). CRI-O also
+needs a `config` with the `mediaType` `application/vnd.oci.image.config.v1+json`,
+while ORAS uses `application/vnd.oci.empty.v1+json` per default. Last but not
+least, CRI-O requires a tarball rather than the plain `.json` file. To achieve
+all of that, the following commands can be executed:
+
+```
+> echo "{}" > config.json
+
+> tar cf seccomp.tar seccomp.json
+
+> oras push \
+    --config config.json:application/vnd.oci.image.config.v1+json \
+     quay.io/crio/seccomp:v1 seccomp.tar
+```
+
+The resulting image contains the right config `mediaType` and pushes a single
+layer `seccomp.tar` to the registry. The name of the profile as well as the
+tarball does not matter much. CRI-O will pick the first `.json` file from any
+OCI artifact and check if that can act as a seccomp profile.
+
+## Future work
+
+CRI-O internally manages the OCI artifacts like regular container images.
+This provides the benefit of moving them around, removing them if not used any
+more or having any other data available than seccomp profiles. This enables future
+enhancements in CRI-O on top of OCI artifacts, but also allows thinking about
+stacking seccomp profiles as part of having multiple layers in an OCI artifact.
+The limitation that it only works for `Unconfined` workloads right now is
```

Comment:
'right now' should be replaced with a release version

Reply:
Yep, done :+1: